build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -1890,6 +1890,71 @@ message DigestFunction {
+    // 
+    // So
+    //
+    //     Hash(blob) = 0x62<git sha1>
+    //     Hash(tree) = 0x74<git sha1>    
+    //
+    // Blob Upload
+    //
+    // A blob is uploaded to the remote side by passing its raw content as well as its
+    // Digest containing the git hash value for a blob prefixed by 0x62.
+    //
+    // Tree Upload
+    //
+    // Whenever a Digest of a Directory message is expected, it can be
+    // convenientely replaced by the Digest of the corresponding git tree.
+    //
+    // Using this git tree representation makes tree handling much more
+    // efficient since the effort of traversing and uploading the content of a
+    // git tree occurs only once and for each subsequent request, and the git
+    // tree id can be directly passed around. We require the invariant that if a
+    // tree is part of any CAS, then all its content is also available in this
+    // CAS. To adhere to this invariant, the client side has to prove that the
+    // content of a tree is available in the CAS before uploading this tree
+    // (i.e., leaves must be uploaded before the root). The server side has to
+    // ensure this invariant holds. In particular, if the remote side implements
+    // any pruning strategy for the CAS, it has to honor this invariant when an
+    // element gets pruned.
+    //
+    // Another consequence of this efficient tree handling is that it improves action
+    // digest calculation noticeably, since known git trees referred by the root
+    // directory do not need to be traversed. This in turn allows to faster determine
+    // whether an action result is already available in the action cache or not.
+    //
+    // Tree Download
+    //
+    // Once an action is successfully executed, it might have generated output files or
+    // output directories in its staging area on the remote side. Each output file
+    // needs to be uploaded to its CAS with the corresponding git blob hash. Each
+    // output directory needs to be translated to a git tree object and uploaded to the
+    // CAS with the corresponding git tree hash. Only if the content of a tree is
+    // available in the CAS is the server side allowed to return the tree to the
+    // client.
+    //
+    // In case of a generated output directory, the server only returns the
+    // corresponding git tree id to the client instead of a flat list of all
+    // recursively generated output directories as part of a Tree Protobuf
+    // message. The remote side promises that each blob and subtree contained in
+    // the root tree is available in the remote CAS. Such blobs and trees must
+    // be accessible, using the streaming interface, without specifying the size
+    // (since sizes are not stored in a git tree). Due to the Protobuf 3
+    // specification, which is used in this remote execution API, not specifying
+    // the size means the default value 0 is used.
+    GITSHA1 = 10;
```

Comment:
So if I understand it correctly, `OutputDirectory.tree_digest` will now be used in two different ways:

1. In case `digest_function != GITSHA1`, it will point to a single REv2 Tree object, which has a height of at most 1.
2. In case `digest_function == GITSHA1`, it will point to a single Git Tree object, which is the equivalent of an REv2 Directory object. It can thus point to a hierarchy of CAS objects whose height is unconstrained.

In my opinion this is a feature that is independent of supporting Git's hashing algorithm and directory format. In #229 we essentially redefined REv2 Tree objects as being a simple binary container format around a sequence of directories. It is generic enough that you could also use that to pack multiple Git directory objects into a single CAS object.

But this is obviously not what you're interested in. You're interested in actually having them stored as separate CAS objects, so that it's possible to embed an output directory into the input root of a subsequent action in a fully opaque manner. And this is a valid request. I think that this is one of the shortcomings of the protocol as stands.

Could you please take a look at PR #258? In that PR I'm proposing that we add a new enumeration to the Command message that clients can use to specify the format in which they want workers to store output directories. By letting Justbuild set that to `DIRECTORY_ONLY`, it can access output directories in Git's native format.

Reply:
Indeed, besides the desire to make use of the fact that `git` naturally has files and directories prehashed, the other desire is to avoid communication overhead in case the client only needs very few artifacts locally (say, the logs of failed tests) and otherwise wants to leave everything on the remote-execution side as it is only used as inputs for other actions. The latter can well be seen as an independent feature that also works with other hashes and directory representations.